File: Excel/900-999/950/950 Challenge.py
import pandas as pd

path = "900-999/950/950 Containers Allotment.xlsx"
input = pd.read_excel(path, usecols="A:B", nrows=21)
test = pd.read_excel(path, usecols="C", nrows=21)

def assign_containers(sizes, capacity=100):
    bins = []
    def place(x):
        for i, used in enumerate(bins):
            if used + x <= capacity:
                bins[i] += x
                return f"C{i+1}"
        bins.append(x)
        return f"C{len(bins)}"
    return [place(x) for x in sizes]

sizes = input['Size'].tolist()
result = assign_containers(sizes)

print(result == test['Answer Expected'].tolist())
# Output: True

File: Excel/900-999/950/950 Challenge.R
library(tidyverse)
library(readxl)

path <- "900-999/950/950 Containers Allotment.xlsx"
input <- read_excel(path, range = "A1:B21")
test <- read_excel(path, range = "C1:C21")

assign_container <- function(sizes, capacity = 100) {
  bins <- numeric()
  map_chr(sizes, function(x) {
    idx <- which(bins + x <= capacity)[1]
    if (is.na(idx)) {
      bins <<- c(bins, x)
      idx <- length(bins)
    } else {
      bins[idx] <<- bins[idx] + x
    }
    paste0("C", idx)
  })
}
result <- input %>%
  mutate(Container = assign_container(Size))

all.equal(result$Container, test$`Answer Expected`)
# [1] TRUE
